### ISSUE TITLE
Improving Accuracy of NBEATS when used on real data

### ISSUE THREAD

**Issue by martin-turion** [unit8co/darts#934]
Hello,

I am attempting to use NBEATS to do forecasting on financial stock tickers.  I just can't seem to get my model to do any sort of accurate forecasting. If you can assist me in letting me know where I am wrong in this process, that would be much appreciated. 

The file I am using is the SPY index from the beginning of time. I will attach it for your own personal use. 
[SPY_formatted.csv](https://github.com/unit8co/darts/files/8597986/SPY_formatted.csv)

### Create timeseries
```python
import pandas as pd
from darts import TimeSeries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from darts.dataprocessing.transformers import Scaler, MissingValuesFiller
from darts.models import NBEATSModel # NN 
from darts.metrics import mape, r2_score

scaler = Scaler()

# Read a pandas DataFrame
ticker_df = pd.read_csv('/content/SPY_formatted.csv', delimiter=",")

# Create a TimeSeries, specifying the time and value columns
ticker_series = TimeSeries.from_dataframe(ticker_df, value_cols='Close')
train, val = ticker_series.split_before(0.90)
train.plot(label="train")
val.plot(label="val")
```
![image](https://user-images.githubusercontent.com/90222598/166125904-5e7277c5-decd-4f9d-a213-f6d94a4f153a.png)

```python
# Scaling Ticker from 0-1 for better training
scaled_ticker_series = scaler.fit_transform(
    filler.transform(
        TimeSeries.from_dataframe(
            ticker_df, value_cols=['Close']
        )
    )
).astype(np.float32)

train_scaled, val_scaled = scaled_ticker_series.split_before(0.90)
train_scaled.plot(label="train")
val_scaled.plot(label="val")
```
![image](https://user-images.githubusercontent.com/90222598/166125897-25d272ef-ca3f-454f-a8e3-5eb9346fc190.png)

### Train model
```python
model_nbeats = NBEATSModel(
    input_chunk_length=30,
    output_chunk_length=7,
    generic_architecture=True,
    num_stacks=10,
    num_blocks=1,
    num_layers=4,
    layer_widths=512,
    n_epochs=100,
    nr_epochs_val_period=1,
    batch_size=800,
    model_name="nbeats_run",
    force_reset=True,
    pl_trainer_kwargs={
      "accelerator": "gpu",
      "gpus": [0]
    },
    save_checkpoints=True,
)

model.fit([train_scaled], val_series= val_scaled)
```
```
Epoch 99: 100%
228/228 [28:18<00:00, 7.45s/it, loss=7.63e-05, train_loss=7.4e-5, val_loss=0.0499]
```

### Prediction

```python
prediction = model.predict(len(val_scaled), series = train_scaled) # predict N number of points given a specific time series
prediction = scaler.inverse_transform(prediction)

ticker_series = scaler.inverse_transform(scaled_ticker_series)
ticker_series.plot()

prediction.plot(label='forecast', low_quantile=0.05, high_quantile=0.95)
plt.legend()
```
![image](https://user-images.githubusercontent.com/90222598/166125949-bce03fbc-e3ae-4483-b75b-208f968699f4.png)

Is there something I am missing? Please let me know if you have any recommendations 😄 
Many thanks in advance.

**Comment by dennisbader** [comment]
Hi @martin-inpel, you are trying to forecast the validation time series in one shot. From your input data those are around 800 predicted target time steps (forecast horizon `n`). Your `output_chunk_length` is 7, so your model was only optimzed to "accurately" predict 7 upcoming time steps. 

When doing one-shot predictions with `n` > output_chunk_length, the models predict auto regressively by consuming its own prior predicitons as input for the new predictions. In each prediction step, it only predicts the next `output_chunk_length=7` time steps.

I would suggest you use methods `model.historical_forecasts()` (see the [N-BEATS example](https://unit8co.github.io/darts/examples/07-NBEATS-examples.html#Generic-architecture)) or `model.backtest()` which consumes the actual data as inputs for all predictions.

Also the SPY index seems to be on a strong positive trend after time step 4000. You're model was trained on all time steps, including the ones before 4000 which might result in lower predicted target values on this steep segment. With parameter `max_samples_per_ts` (when calling `model.fit()`) you can limit the training samples to only take in the last `max_samples_per_ts` before the end of your training series. This might lead to better performance.

Hope this helped.

**Comment by martin-inpel** [comment]
Hi @dennisbader , thanks for the great information! 

Few simple follow up questions: 
1.  How does `max_samples_per_ts ` differ from `batch_size`. Wouldnt that be the same thing when training? 
  i) how would you set the `max_samples_per_ts` for this dataset if I were looking to do varing forcasts of 7 days OR 30 days.  
3. So to increase accuracy over periods longer than 7, I should be training with greater `output_chunk_length`. Any rule of thumb on what to make `input_chuck_length`? 3 times the size?

**Comment by hrzn** [comment]
> Hi @dennisbader , thanks for the great information!
> 
> Few simple follow up questions:
> 
> 1. How does `max_samples_per_ts ` differ from `batch_size`. Wouldnt that be the same thing when training?

`max_samples_per_ts` refers to the number of training samples (input/output examples) that are used to train the model. `batch_size` refers to the batch size - i.e., how many of those training samples are batched together in a forward pass of the model.

>    i) how would you set the `max_samples_per_ts` for this dataset if I were looking to do varing forcasts of 7 days OR 30 days.

You have to experiment. Lower values of `max_samples_per_ts` will use more recent samples only.

> 2. So to increase accuracy over periods longer than 7, I should be training with greater `output_chunk_length`. Any rule of thumb on what to make `input_chuck_length`? 3 times the size?

Here too this is a hyper-parameter and you have to experiment, but trying multiples of the output length makes sense as a rule of thumb, IMO.


PS: using forecasting models to predict stock prices based only on their history seems like an audacious endeavour :D It's a good learning exercise, though.

**Comment by martin-inpel** [comment]
@hrzn  Thank you for the information.

> PS: using forecasting models to predict stock prices based only on their history seems like an audacious endeavour :D It's a good learning exercise, though.

Yes, I know... It's fun to use real world examples, even though the markets make no sense right now 😄 

@hrzn  For further clarification: 
`input_chunk_length` should equal to `max_samples_per_ts` to train the models correct? 
therefore, for my example, 
`input_chunk_length` = 30
`max_samples_per_ts` = 30

Would this be correct in my thinking?
